vm-virtio/src/mem.rs

```diff
@@ -0,0 +1,951 @@
+
+    state: VirtioMemRespState,
+}
+
+// Safe because it only has data and has no implicit padding.
+unsafe impl ByteValued for VirtioMemResp {}
+
+/// Got from qemu/include/standard-headers/linux/virtio_mem.h
+#[repr(C, packed)]
+#[derive(Copy, Clone, Debug, Default)]
+struct VirtioMemConfig {
+    /* Block size and alignment. Cannot change. */
+    block_size: u32,
+    /* Valid with VIRTIO_MEM_F_ACPI_PXM. Cannot change. */
+    node_id: u16,
+    padding: u16,
+    /* Start address of the memory region. Cannot change. */
+    addr: u64,
+    /* Region size (maximum). Cannot change. */
+    region_size: u64,
+    /*
+     * Currently usable region size. Can grow up to region_size. Can
+     * shrink due to VIRTIO_MEM_REQ_UNPLUG_ALL (in which case no config
+     * update will be sent).
+     */
+    usable_region_size: u64,
+    /*
+     * Currently used size. Changes due to plug/unplug requests, but no
+     * config updates will be sent.
+     */
+    plugged_size: u64,
+    /* Requested size. New plug requests cannot exceed it. Can change. */
+    requested_size: u64,
+}
+
+// Safe because it only has data and has no implicit padding.
+unsafe impl ByteValued for VirtioMemConfig {}
+
+struct Request {
+    req: VirtioMemReq,
+    status_addr: GuestAddress,
+}
+
+impl Request {
+    fn parse(
+        avail_desc: &DescriptorChain,
+        mem: &GuestMemoryMmap,
+    ) -> result::Result<Request, Error> {
+        // The head contains the request type which MUST be readable.
+        if avail_desc.is_write_only() {
+            return Err(Error::UnexpectedWriteOnlyDescriptor);
+        }
+        if avail_desc.len as usize != size_of::<VirtioMemReq>() {
+            return Err(Error::InvalidRequest);
+        }
+        let req: VirtioMemReq = mem.read_obj(avail_desc.addr).map_err(Error::GuestMemory)?;
+
+        let status_desc = avail_desc
+            .next_descriptor()
+            .ok_or(Error::DescriptorChainTooShort)?;
+
+        // The status MUST always be writable
+        if !status_desc.is_write_only() {
+            return Err(Error::UnexpectedReadOnlyDescriptor);
+        }
+
+        if (status_desc.len as usize) < size_of::<VirtioMemResp>() {
+            return Err(Error::BufferLengthTooSmall);
+        }
+
+        Ok(Request {
+            req,
+            status_addr: status_desc.addr,
+        })
+    }
+}
+
+pub struct Resize {
+    size: Arc<AtomicU64>,
+    tx: mpsc::Sender<Result<(), Error>>,
+    rx: Option<mpsc::Receiver<Result<(), Error>>>,
+    evt: EventFd,
+}
+
+impl Resize {
+    pub fn new() -> io::Result<Self> {
+        let (tx, rx) = mpsc::channel();
+
+        Ok(Resize {
+            size: Arc::new(AtomicU64::new(0)),
+            tx,
+            rx: Some(rx),
+            evt: EventFd::new(EFD_NONBLOCK)?,
+        })
+    }
+
+    pub fn try_clone(&self) -> Result<Self, Error> {
+        Ok(Resize {
+            size: self.size.clone(),
+            tx: self.tx.clone(),
+            rx: None,
```

Comment:
Channel receiver cannot be cloned because a channel allows only one receiver.
And for the current code, the VMM should keep the receiver to get the reply from virtio-mem device.  So rx don't need to be cloned.